x/ccv/provider/keeper/key_assignment.go

```diff
@@ -0,0 +1,618 @@
+			)
+		} else {
+			// the validator had no key assigned on this consumer chain
+			providerKey, err := validator.TmConsPublicKey()
+			if err != nil {
+				return err
+			}
+			oldConsumerKey = providerKey
+		}
+
+		// check whether the validator is valid, i.e., its power is positive
+		oldPower := k.stakingKeeper.GetLastValidatorPower(ctx, sdk.ValAddress(validator.OperatorAddress))
+		if oldPower > 0 {
+			// to enable multiple calls of AssignConsumerKey in the same block by the same validator
+			// the pending key assignment should not be overwritten
+			if _, found := k.GetKeyAssignmentReplacement(ctx, chainID, providerAddr); !found {
+				// store old key and power for modifying the valset update in EndBlock;
+				// note: this state is deleted at the end of the block
+				k.SetKeyAssignmentReplacement(
+					ctx,
+					chainID,
+					providerAddr,
+					abci.ValidatorUpdate{PubKey: oldConsumerKey, Power: oldPower},
+				)
+			}
+		}
+	} else {
+		// if the consumer chain is not registered, then remove the mapping
+		// from the old consumer address to the provider address (if any)
+		// get the previous key assigned for this validator on this consumer chain
+		oldConsumerKey, found := k.GetValidatorConsumerPubKey(ctx, chainID, providerAddr)
+		if found {
+			k.DeleteValidatorByConsumerAddr(
+				ctx,
+				chainID,
+				utils.TMCryptoPublicKeyToConsAddr(oldConsumerKey),
+			)
+		}
+	}
+
+	// set the mapping from this validator's provider address to the new consumer key;
+	// overwrite if already exists
+	// note: this state is deleted when the validator is removed from the staking module
+	k.SetValidatorConsumerPubKey(ctx, chainID, providerAddr, consumerKey)
+
+	// set the mapping from this validator's new consensus address on the consumer
+	// to its consensus address on the provider;
+	// note: this state must be deleted through the pruning mechanism
+	k.SetValidatorByConsumerAddr(ctx, chainID, consumerAddr, providerAddr)
+
+	return nil
+}
+
+// ApplyKeyAssignmentToInitialValset applies the key assignment to the initial validator
+// for a consumer chain
+func (k Keeper) ApplyKeyAssignmentToInitialValset(
+	ctx sdk.Context,
+	chainID string,
+	updates []abci.ValidatorUpdate,
+) (newUpdates []abci.ValidatorUpdate) {
+	newUpdates = updates
+	for i, update := range updates {
+		providerAddr := utils.TMCryptoPublicKeyToConsAddr(update.PubKey)
+		if consumerKey, found := k.GetValidatorConsumerPubKey(ctx, chainID, providerAddr); found {
+			newUpdates[i].PubKey = consumerKey
+		}
+	}
+	return newUpdates
+}
+
+// ApplyKeyAssignmentToValUpdates applies the key assignment to the validator updates
+// received from the staking module
+func (k Keeper) ApplyKeyAssignmentToValUpdates(
+	ctx sdk.Context,
+	chainID string,
+	valUpdates []abci.ValidatorUpdate,
+) (newUpdates []abci.ValidatorUpdate, err error) {
+	for _, valUpdate := range valUpdates {
+		providerAddr := utils.TMCryptoPublicKeyToConsAddr(valUpdate.PubKey)
```

Comment:
If the provider did not assign a consumer key before the chain started and the chain has been running for some time, we also need to set the validator power to 0 for the provider's "normal" key.

Otherwise, the "normal" key stays in the validator set and the consumer key is added to the list.

From integration tests:
```bash
$: interchain-security-pd q tendermint-validator-set --node tcp://7.7.7.253:26658

block_height: "91"
total: "3"
validators:
// this is  the  "normal" key of the provider
- address: cosmosvalcons1qmq08eruchr5sf5s3rwz7djpr5a25f7xw4mceq
  proposer_priority: "-758"
  pub_key:
    type: tendermint/PubKeyEd25519
    value: RrclQz9bIhkIy/gfL485g3PYMeiIku4qeo495787X10=
  voting_power: "511"
- address: cosmosvalcons1nx7n5uh0ztxsynn4sje6eyq2ud6rc6klc96w39
  proposer_priority: "379"
  pub_key:
    type: tendermint/PubKeyEd25519
    value: mAN6RXYxSM4MNGSIriYiS7pHuwAcOHDQAy9/wnlSzOI=
  voting_power: "500"
- address: cosmosvalcons1ezyrq65s3gshhx5585w6mpusq3xsj3ayzf4uv6
  proposer_priority: "379"
  pub_key:
    type: tendermint/PubKeyEd25519
    value: IHo4QEikWZfIKmM0X+N+BjKttz8HOzGs2npyjiba3Xk=
  voting_power: "500"


// check consumer valset after key assignment
$: interchain-security-cd q tendermint-validator-set --node tcp://7.7.8.253:26658

block_height: "65"
total: "4"
validators:
// NEW KEY -> pubkey assigned by provider
- address: cosmosvalcons16q4jna57d07rufql8qtzaqhwf3ud3hdf5yt3y2
  proposer_priority: "480"
  pub_key:
    type: tendermint/PubKeyEd25519
    value: GJuUXISPjcWRIbEdzLTtVHzhnt9T98URH/gOA8KB7fA=
  voting_power: "511"
// OLD KEY -> the  "normal" key of the provider
- address: cosmosvalcons1qmq08eruchr5sf5s3rwz7djpr5a25f7xw4mceq
  proposer_priority: "-1500"
  pub_key:
    type: tendermint/PubKeyEd25519
    value: RrclQz9bIhkIy/gfL485g3PYMeiIku4qeo495787X10=
  voting_power: "500"
- address: cosmosvalcons1nx7n5uh0ztxsynn4sje6eyq2ud6rc6klc96w39
  proposer_priority: "511"
  pub_key:
    type: tendermint/PubKeyEd25519
    value: mAN6RXYxSM4MNGSIriYiS7pHuwAcOHDQAy9/wnlSzOI=
  voting_power: "500"
- address: cosmosvalcons1ezyrq65s3gshhx5585w6mpusq3xsj3ayzf4uv6
  proposer_priority: "511"
  pub_key:
    type: tendermint/PubKeyEd25519
    value: IHo4QEikWZfIKmM0X+N+BjKttz8HOzGs2npyjiba3Xk=
  voting_power: "500"
```

Reply:
We already do that, see https://github.com/cosmos/interchain-security/blob/c396b3f704842b195a318f5e871fecd3e5b142da/x/ccv/provider/keeper/key_assignment.go#L422

This key is added to the `KeyAssignmentReplacement` here https://github.com/cosmos/interchain-security/blob/c396b3f704842b195a318f5e871fecd3e5b142da/x/ccv/provider/keeper/key_assignment.go#L432 and then used here https://github.com/cosmos/interchain-security/blob/c396b3f704842b195a318f5e871fecd3e5b142da/x/ccv/provider/keeper/key_assignment.go#L483

Can you output the VSCPacket sent?